### PR TITLE
Add Table.parallel_apply

### PR DESCRIPTION
**Pull request by spenczar**
This is a minimalist, simple implementation of multithreaded computation on a quivr Table. 

The Table is sliced using RecordBatches, which don't require a copy. Each thread gets a slice. The threads work on their chunk and yield back arbitrary data which gets fed out as an iterator.

The test shows how it can be used. You can apply a function, a lambda, a closure, or a method - they'll all work.

I have not yet benchmarked this, but I have done very rudimentary checks on memory usage (like, squint at `top`) and think it doesn't do anything too silly.

### CODE REVIEW

File: quivr/tables.py [hunk 1]
```diff
@@ -1082,3 +1085,44 @@ def set_column(self, name: str, data: DataSourceType) -> Self:
...
+        """
+        if max_workers is None:
+            max_workers = os.cpu_count()
+
+        futures = []
+
+        def actual_func(batch: pa.RecordBatch) -> T:
+            table = pa.Table.from_batches([batch], schema=self.table.schema)
+            # Assume the table remains valid for batches
+            instance = self.from_pyarrow(table, validate=False)
```
Comment (akoumjian) [inline quivr/tables.py:1119]: I'm assuming this is basically zero copy with only the tiny overhead of the quivr table instance pointing to the batch.

**Review by akoumjian — APPROVED** [review]
I like how simple it is and I think it's good for us to resist trying to optimize on things like chunking at this point.

I do think there may be times where we use a very similar pattern outside of this method and end up repeating ourselves and that's okay. For instance, it may feel awkward to use closures to send in other sets of batched data. Although I almost most of the time we will only be batching one table and sending in other relational tables in their entirety and do groupby's or filtering inside the individual threads.

The nice thing here is that it shows the proper way to chunk up a pyarrow table row-wise and then reinitialize it into a quivr table without consuming more memory.

**Review by akoumjian — CHANGES_REQUESTED** [review]
Actually, I just realized I don't believe this works for our use case. We can't use threads because we're CPU bound, right?

**Comment by akoumjian** [comment]
I'm afraid we might need something like this (snippet provided by ChatGPT):

```python
import pyarrow as pa
from concurrent.futures import ProcessPoolExecutor
from multiprocessing import shared_memory
import numpy as np

def worker(shm_name, schema):
    # Attach to the existing shared memory block
    shm = shared_memory.SharedMemory(name=shm_name)

    # Create a pyarrow buffer referencing the shared memory
    buffer = pa.py_buffer(shm.buf)
    
    # Create a pyarrow table from the buffer using the schema
    reader = pa.RecordBatchFileReader(buffer)
    table = pa.Table.from_batches([reader.get_batch(0)])
    
    # Perform some CPU-intensive task on the table
    # ...
    
    # Cleanup: unlink the shared memory block (this doesn't remove it)
    shm.unlink()

    return "some result"

if __name__ == "__main__":
    # Define a schema for your table (adjust as necessary)
    schema = pa.schema([("field1", pa.int32()), ("field2", pa.string())])
    
    # Create a pyarrow table (or obtain it in some other manner)
    data1 = np.arange(1000, dtype=np.int32)
    data2 = np.array([str(i) for i in range(1000)], dtype=object)
    
    table = pa.Table.from_arrays([pa.array(data1), pa.array(data2)], schema=schema)
    
    # Create a shared memory block and write the pyarrow table's data to it
    shm = shared_memory.SharedMemory(create=True, size=table.nbytes)
    
    # Write the table's record batches to the shared memory block
    with pa.RecordBatchFileWriter(shm.buf, schema) as writer:
        writer.write_table(table)
    
    # Start a process pool and pass the shared memory name to the worker
    with ProcessPoolExecutor() as executor:
        future = executor.submit(worker, shm.name, schema)
        
        # Get the result from the worker
        result = future.result()
        print("Worker returned:", result)
    
    # Cleanup: close and unlink the shared memory block
    shm.close()
    shm.unlink()
```

**Comment by spenczar** [comment]
> Actually, I just realized I don't believe this works for our use case. We can't use threads because we're CPU bound, right?

I don't understand what you mean here. With one thread per core, this code would use all CPU cores, and I don't think there is any cross talk other than the main thread collecting results. Why would processes be different?

**Comment by moeyensj** [comment]
This is pretty sweet!

One operation we will be doing a lot is group-by then apply_parallel (group-by orbit ID, then apply iod function; group-by observation times then apply range and transform, etc...). This might be out-of-scope for this but is there a pathway forward towards something like that?

**Comment by spenczar** [comment]
Ah, right, duh. Pyarrow.compute doesn't need the GIL, and same with many numpy operations, but yes - we probably _do_ need to drop down to multiprocessing.

FWIW, the issue with the code that GPT spat out is that it puts the burden of reference counting on the caller; a failure to call `shm.unlink` can leak, but also can make the caller's `shm.close` block indefinitely. I have a branch that tries to do it all correctly; I'll resuscitate that thing.

**Comment by spenczar** [comment]
See https://github.com/spenczar/quivr/pull/50.

**Comment by spenczar** [comment]
Replaced with #50.

**Pull request closed without merging**